service/matching/matching_engine.go

```diff
@@ -1113,7 +1113,13 @@ func (e *matchingEngineImpl) GetTaskQueueUserData(
 	for {
 		resp := &matchingservice.GetTaskQueueUserDataResponse{}
 		userData, userDataChanged, err := tqMgr.GetUserData()
-		if err != nil {
+		if err == errTaskQueueClosed {
+			// If we're closing, return a success with no data, as if the request expired. We shouldn't
+			// close due to idleness (because of the MarkAlive above), so we're probably closing due to a
+			// change of ownership. The caller will retry and be redirected to the new owner.
+			resp.TaskQueueHasUserData = userData != nil
```

Comment:
resp.TaskQueueHasUserData = true AND resp.UserData = nil seems like a bug. Do we need a better more explicit contract here for caller to retry?

Reply:
It is intentional: `TaskQueueHasUserData` means "is there any user data set on this task queue", `UserData` not being nil means "the user data has changed since the version you have, here's the new data"

However, it turns out no one ever uses `TaskQueueHasUserData`, and I have another PR (not sent yet) to delete it. So it doesn't matter.